overlord/snapstate/snapmgr.go

```diff
@@ -738,13 +748,19 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	if err != nil {
 		return err
 	}
+	var oldCurrent snap.Revision
+	err = t.Get("old-current", &oldCurrent)
+	if err != nil {
+		return err
+	}
 
 	// relinking of the old snap is done in the undo of unlink-current-snap
 
 	snapst.Candidate = snapst.Sequence[len(snapst.Sequence)-1]
 	snapst.Sequence = snapst.Sequence[:len(snapst.Sequence)-1]
 	snapst.Active = false
 	snapst.Channel = oldChannel
+	snapst.Current = oldCurrent
```

Comment:
would move closer to Sequence actually (we should always pair changing sequence with changing Current), this is a general comment if you can recheck, may not be easy to spot all these cases just from the diff